qa: psalm improvements

gsteel opened the pull request:
|    Q          |   A
|-------------- | ------
| Bugfix        | yes
| BC Break      | no
| QA            | yes

### Description

Trying to make some improvements to `input-filter` has led me down a psalm rabbit hole…

Whilst this pull is a general reduction in the baseline and improvement for AbstactValidator and the plugin manager, it has introduced some new errors that I've baselined because I don't have the time - most of them appear to be constructor incompatibilities in the many validators - i.e. annotating each individual constructor with `@param array<string, mixed>|Traversable<string, mixed> $options`

Spotted a couple of void returns when they were declared as whatever|null

I'll do a review and point out as much as possible - Hopefully, on balance this is an improvement…

gsteel committed:
qa: psalm improvements

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented on src/AbstractValidator.php:
```diff
@@ -26,6 +26,20 @@
+ *     translatorEnabled: bool,
+ *     valueObscured: bool,
+ * }
+ * @property array<string, mixed> $options
+ * @property array<string, string> $messageTemplates
+ * @property array<string, mixed> $messageVariables
```
declaring these magic properties has caused an increase in psalm errors that should be easily fixed in individual validators

gsteel commented on src/AbstractValidator.php:
```diff
@@ -296,29 +322,29 @@ public function __get($property)
     protected function createMessage($messageKey, $value)
     {
         if (! isset($this->abstractOptions['messageTemplates'][$messageKey])) {
-            return;
+            return null;
```
This was a bug right?

froschdesign replied:
Very old remains.

gsteel commented on src/AbstractValidator.php:
```diff
@@ -351,11 +377,19 @@ protected function error($messageKey, $value = null)
         }
 
-        $this->abstractOptions['messages'][$messageKey] = $this->createMessage($messageKey, $value);
+        $message = $this->createMessage($messageKey, $value);
+        if (! is_string($message)) {
+            return;
+        }
+
+        $this->abstractOptions['messages'][$messageKey] = $message;
```
Message is possibly null here, so an early return prevents `null` from being added to the list of messages - perhaps a subtle BC break?

Ocramius replied:
I think it's OK to prevent `null` from leaking in here.

gsteel commented on src/AbstractValidator.php:
```diff
@@ -427,7 +462,7 @@ public function setTranslator(?Translator\TranslatorInterface $translator = null
     public function getTranslator()
     {
         if (! $this->isTranslatorEnabled()) {
-            return;
+            return null;
```
Another void return which _should_ be null

froschdesign replied:
Same here.

gsteel commented on src/ValidatorChain.php:
```diff
@@ -239,12 +244,13 @@ public function isValid($value, $context = null)
         foreach ($this->validators as $element) {
             $validator = $element['instance'];
+            assert($validator instanceof ValidatorInterface);
             if ($validator->isValid($value, $context)) {
                 continue;
             }
             $result         = false;
             $messages       = $validator->getMessages();
-            $this->messages = array_replace_recursive($this->messages, $messages);
+            $this->messages = array_replace($this->messages, $messages);
```
messages is `array<string, string>` therefore `array_replace` is sufficient and retains the existing type for psalm

Ocramius commented on src/AbstractValidator.php:
```diff
@@ -26,6 +26,20 @@
+ *     translatorEnabled: bool,
+ *     valueObscured: bool,
+ * }
+ * @property array<string, mixed> $options
```
Are these `public`? Because `@property` makes them `public` :thinking:

gsteel replied:
I guess if they're undeclared they are already public no? Typically they'd be protected in extending classes, if set at all, so I guess not. I'm unsure as to why `$abstractOptions` exists at all and why these properties aren't just declared on the abstract

All I actually wanted to achieve was to nail down what `getMessages()` returns but doing that is a bit of an endless battle.

Ocramius replied:
Let's keep it as-is then: still a good improvement :+1:

gsteel committed:
Drop the added assertion that affects the mime-type test

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/AbstractValidator.php:
```diff
@@ -78,20 +91,23 @@ abstract class AbstractValidator implements
         // The abstract constructor allows no scalar values
         if ($options instanceof Traversable) {
+            /** @psalm-var array<string, mixed> $options */
             $options = ArrayUtils::iteratorToArray($options);
```
TBH, sounds like we could improve inference here, in future

gsteel replied:
I'll send in a pull for `ArrayUtils::iteratorToArray` asap

gsteel replied:
https://github.com/laminas/laminas-stdlib/pull/62

Ocramius replied:
@gsteel released!

gsteel committed:
Update stdlib to 3.10 and remove unnecessary annotation

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on .psr-container-stub.php:
```diff
@@ -0,0 +1,21 @@
+<?php
```
Following discussion at https://github.com/Lctrs/psalm-psr-container-plugin/issues/545#issuecomment-1150465946, can we:

 1. export-ignore this
 2. perhaps rename it to `.psr-container-stub.php.stub`?

Ocramius replied:
We would probably drop a line of comment explaining why this doesn't end with `.php` :D

gsteel replied:
👍 - Had a general tidy-up of the dot-files/ignores etc too and got rid of coveralls

Ocramius replied:
Indeed, awesome!

Ocramius reviewed (approved):
LGTM! Waiting for @gsteel's feedback and potential changes on the stub file, before merging

gsteel committed:
Re-organize dot-files

- Rename Psr\Container stub to avoid detection as a duplicate interface
- Update export-ignores in .gitattributes to avoid shipping unnecessary files
- Remove coveralls badge and assets as they are no longer in use

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented:
Thanks @gsteel!
